doc/arif_rime.3
.TH ARIF_RIME 3 "Jan 02, 2023" 0.1.0 "ARIF User Manual"
.
.SH NAME
arif_rime - Rime IME integration for ARIF
.
.SH SYNOPSIS
.EX
.B #include <arif_rime.h>
.PP
.BI "extern struct arif_engine const " arif_rime_engine ;
.EE
.
.SH DESCRIPTION
Example implementation of an ARIF engine, powered by the Rime IME core library.
.PP
See the manual page of
.BR arif_set_engine (3)
for how to use this engine.
.SS Initialization
The
.IR init ()
function of
.I arif_rime_engine
ignores the first argument.
Engine configuration is instead read from environment variables.
.PP
Note that the variables are only read during the first
.IR init ()
call, initialization of subsequent engine instances share the same
configuration.
.
.SH ENVIRONMENT
.TP
.B ARIF_RIME_USER_DATA_DIR
Path to the directory where Rime user data files are stored.
.IP
User data can be initialized with
.BR rime_deployer (1).
.TP
.B ARIF_RIME_SHARED_DATA_DIR
Path to the directory where Rime schema data files are stored.
.IP
Defaults to "/usr/share/rime\-data".
.TP
.B ARIF_RIME_LOG_DIR
Path to the directory where Rime log files are written to.
.IP
Defaults to "/tmp".
.TP
.B ARIF_RIME_LOG_LEVEL
Rime minimum log level.
Should be one of INFO, WARNING, ERROR or FATAL (case insensitive).
.IP
Default log level is ERROR.
.TP
.B ARIF_RIME_MODULES
A comma-separated list of Rime modules to load before Rime initialization.
.
.SH BUGS
This is only an example to shed light on how an ARIF engine can be implemented.
.PP
It is not intended for everyday use (it could be, depending on the use cases),
since only part of Rime's features are supported, and may not work as expected
with certain configurations.
.
.SH COPYRIGHT
Copyright (C) 2023,2024  CismonX <admin@cismon.net>
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.
.PP
You should have received a copy of the license along with this document.
If not, see
.UR https://www.gnu.org/licenses/fdl-1.3.html
.UE .
.
.SH SEE ALSO
.BR arif_set_engine (3)
.PP
Rime IME core library:
.UR https://github.com/rime/librime
.UE


src/arif_rime.c
/**
 * arif/src/arif_rime.c - Rime IME integration for ARIF
 * ----
 *
 * Copyright (C) 2023  CismonX <admin@cismon.net>
 *
 * This file is part of ARIF, Another Readline Input Framework.
 *
 * ARIF is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ARIF is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ARIF.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include "arif_rime.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <strings.h>

#include <rime_api.h>

#include "arif_common.h"

struct engine_ctx {
    RimeSessionId      session;
    struct cand_entry *candidates;
};

// Forward declaration start
static void          arif_rime_finalize (void *);
static void          arif_rime_info     (void *, char const **, char const **);
static int           arif_rime_init     (void *, void **);
static int           arif_rime_query    (void *, char const *, int, int, int,
                                         struct arif_cand const **);
static int           copy_candidate     (struct arif_cand *, RimeContext *,
                                         char const *, int);
static int           init_rime          (void);
static void          finalize_rime      (void);
static void          free_candidates    (struct engine_ctx *);
static char const *  gen_description    (void);
static int           get_log_level      (void);
static char const ** get_modules        (void);
// Forward declaration end

struct cand_entry {
    struct cand_entry *next;
    int                num_candidates;
    struct arif_cand   candidates[];
};

struct arif_engine const arif_rime_engine = {
    .init     = arif_rime_init,
    .finalize = arif_rime_finalize,
    .info     = arif_rime_info,
    .query    = arif_rime_query,
};

static RimeApi *rime_api;
static char    *rime_description;

// XXX: glog (used by Rime) may break Readline output by writing to stderr.
//      This is a hack to override and disable the corresponding flags
//      without adding glog and a C++ compiler as our dependency.
#ifndef ARIF_RIME_GLOG_ALSOLOGTOSTDERR
#  define ARIF_RIME_GLOG_ALSOLOGTOSTDERR  _ZN3fLB21FLAGS_alsologtostderrE
#endif
#ifndef ARIF_RIME_GLOG_STDERRTHRESHOLD
#  define ARIF_RIME_GLOG_STDERRTHRESHOLD  _ZN3fLI21FLAGS_stderrthresholdE
#endif

int ARIF_RIME_GLOG_ALSOLOGTOSTDERR;  // fLB::FLAGS_alsologtostderr
int ARIF_RIME_GLOG_STDERRTHRESHOLD;  // fLI::FLAGS_stderrthreshold

static void
arif_rime_finalize (
    void *engine_data
) {
    struct engine_ctx *ctx = engine_data;

    free_candidates(ctx);
    if (rime_api != NULL) {
        rime_api->destroy_session(ctx->session);
    }
    free(ctx);
}

static void
arif_rime_info (
    void        *ARIF_UNUSED_ARG(engine_data),
    char const **name_ptr,
    char const **description_ptr
) {
    if (name_ptr != NULL) {
        *name_ptr = "rime";
    }
    if (description_ptr != NULL) {
        *description_ptr = gen_description();
    }
}

static int
arif_rime_init (
    void  *ARIF_UNUSED_ARG(args),
    void **engine_data_ptr
) {
    if (0 != init_rime()) {
        return -1;
    }

    struct engine_ctx *ctx = xmalloc(sizeof(struct engine_ctx));
    *ctx = (struct engine_ctx) {
        .session = 0,
    };
    *engine_data_ptr = ctx;
    return 0;
}

static int
arif_rime_query (
    void                    *engine_data,
    char const              *line,
    int                      offset,
    int                      len,
    int                      num_candidates,
    struct arif_cand const **candidates_ptr
) {
    struct engine_ctx *ctx = engine_data;

    char *prefix     = NULL;
    int   prefix_len = 0;

    if (line != NULL) {
        free_candidates(ctx);
        rime_api->destroy_session(ctx->session);
        ctx->session = rime_api->create_session();

        char const *text = line + offset;
        for (int idx = 0; idx < len; ++idx) {
            RIME_STRUCT(RimeStatus, rimestatus);
            if (!rime_api->process_key(ctx->session, text[idx], 0) ||
                    !rime_api->get_status(ctx->session, &rimestatus)
            ) {
                free(prefix);
                return 0;
            }

            int is_composing = rimestatus.is_composing;
            rime_api->free_status(&rimestatus);
            if (is_composing) {
                continue;
            }

            // When `is_composing == false`, it means that
            // the previous input has triggered a commit.
            RIME_STRUCT(RimeCommit, rimecommit);
            if (!rime_api->get_commit(ctx->session, &rimecommit)) {
                continue;
            }
            int commit_len = strlen(rimecommit.text);
            prefix = xrealloc(prefix, prefix_len + commit_len);
            memcpy(prefix + prefix_len, rimecommit.text, commit_len);
            prefix_len += commit_len;
            rime_api->free_commit(&rimecommit);
        }
    } else if (candidates_ptr == NULL) {
        rime_api->select_candidate(ctx->session, offset);
        return 0;
    } else {
        if (!rime_api->find_session(ctx->session)) {
            return 0;
        }
    }

    size_t candidates_size = sizeof(struct arif_cand) * num_candidates;
    struct cand_entry *entries
            = xmalloc(sizeof(struct cand_entry) + candidates_size);
    entries->next = ctx->candidates;
    ctx->candidates = entries;

    int idx = 0;
    while (idx < num_candidates) {
        RIME_STRUCT(RimeContext, rimectx);
        if (!rime_api->get_context(ctx->session, &rimectx)) {
            break;
        }

        struct arif_cand *candidate = entries->candidates + idx;
        int result = copy_candidate(candidate, &rimectx, prefix, prefix_len);

        rime_api->free_context(&rimectx);

        if (result < 0) {
            // no candidates
            break;
        }
        ++idx;
        if (result > 0) {
            // last candidate
            break;
        }
        // 0xff54 == XK_Down (see X11/keysyms.h)
        if (!rime_api->process_key(ctx->session, 0xff54, 0)) {
            break;
        }
    }

    if (idx == 0 && prefix_len > 0) {
        struct arif_cand *cand = &entries->candidates[0];

        *cand = (struct arif_cand) {
            .text        = prefix,
            .text_len    = prefix_len,
            .replace_len = len,
        };
        idx = 1;
    } else {
        free(prefix);
    }

    entries->num_candidates = idx;
    *candidates_ptr = entries->candidates;
    return idx;
}

static int
copy_candidate (
    struct arif_cand *dest,
    RimeContext      *src,
    char const       *prefix,
    int               prefix_len
) {
    RimeMenu *menu = &src->menu;
    int candidate_idx = menu->highlighted_candidate_index;
    if (candidate_idx >= menu->page_size) {
        return -1;
    }

    RimeCandidate   *candidate   = &menu->candidates[candidate_idx];
    RimeComposition *composition = &src->composition;

    char const *text = candidate->text;
    if (text == NULL) {
        text = "";
    }
    int text_len = strlen(text);

    if (prefix == NULL) {
        prefix = "";
    }

    char const *comment = candidate->comment;
    if (comment == NULL) {
        comment = "";
    }
    int comment_len = strlen(comment);

    char const *line     = composition->preedit;
    int         line_len = composition->length;

    char *buf = xmalloc(text_len + prefix_len + line_len + comment_len + 1);
    memcpy(buf,                                    text,    text_len);
    memcpy(buf + text_len,                         prefix,  prefix_len);
    memcpy(buf + text_len + prefix_len,            line,    line_len);
    memcpy(buf + text_len + prefix_len + line_len, comment, comment_len);

    // Distinguish "no comment" from "empty comment",
    // so that the frontend can handle them differently.
    if (candidate->comment == NULL) {
        comment = NULL;
    } else {
        comment = buf + text_len + line_len;
    }
    *dest = (struct arif_cand) {
        .text          = buf,
        .text_len      = text_len,
        .replace_start = prefix_len + composition->sel_start,
        .replace_len   = composition->sel_end - composition->sel_start,
        .transform     = buf + text_len,
        .transform_len = prefix_len + line_len,
        .display       = comment,
        .display_len   = comment_len,
    };

    return menu->is_last_page && candidate_idx + 1 == menu->page_size;
}

static void
free_candidates (
    struct engine_ctx *ctx
) {
    struct cand_entry *entry, *next;
    for (entry = ctx->candidates; entry != NULL; entry = next) {
        for (int idx = 0; idx < entry->num_candidates; ++idx) {
            struct arif_cand *candidate = entry->candidates + idx;
            free((char *) candidate->text);
        }

        next = entry->next;
        free(entry);
    }
    ctx->candidates = NULL;
}

static int
init_rime (void)
{
    if (rime_api != NULL) {
        return 0;
    }

    rime_api = rime_get_api();
    if (rime_api == NULL) {
        return -1;
    }

    RIME_STRUCT(RimeTraits, traits);
    traits.app_name               = "rime.arif";
    traits.distribution_name      = "Rime";
    traits.distribution_code_name = "arif-rime";
    traits.distribution_version   = ARIF_VER_STR;
    traits.shared_data_dir
            = xgetenv("ARIF_RIME_SHARED_DATA_DIR", "/usr/share/rime-data");
    traits.user_data_dir = xgetenv("ARIF_RIME_USER_DATA_DIR", NULL);
    traits.log_dir       = xgetenv("ARIF_RIME_LOG_DIR", "/tmp");
    traits.min_log_level = get_log_level();
    traits.modules       = get_modules();
    rime_api->setup(&traits);

    // Rime sets this to true during init
    ARIF_RIME_GLOG_ALSOLOGTOSTDERR = 0;
    // default value is ERROR (2) instead of FATAL
    ARIF_RIME_GLOG_STDERRTHRESHOLD = 3;

    rime_api->initialize(&traits);
    // Rime uses static C++ variables to store its states, which get destroyed
    // on exit.  Make sure we finalize Rime API before that.
    atexit(finalize_rime);

    if (traits.modules != NULL) {
        free((char *) traits.modules[0]);
        free(traits.modules);
    }
    return 0;
}

static void
finalize_rime (void)
{
    if (rime_api != NULL) {
        rime_api->finalize();
        rime_api = NULL;
    }
    free(rime_description);
}

static char const *
gen_description (void)
{
    if (rime_description != NULL) {
        return rime_description;
    }

    char const *rime_version = rime_api->get_version();
    if (rime_version == NULL) {
        rime_version = "N/A";
    }

    char const *fmt = "RIME integration for ARIF\n"
            "  Rime IME <https://rime.im/> version: %s\n";
    int desc_len = snprintf(NULL, 0, fmt, rime_version);
    rime_description = xmalloc(sizeof(char) * (desc_len + 1));
    sprintf(rime_description, fmt, rime_version);
    return rime_description;
}

static int
get_log_level (void)
{
    char const *log_level = xgetenv("ARIF_RIME_LOG_LEVEL", "");

    if (0 == strcasecmp("INFO", log_level)) {
        return 0;
    }
    if (0 == strcasecmp("WARNING", log_level)) {
        return 1;
    }
    if (0 == strcasecmp("ERROR", log_level)) {
        return 2;
    }
    if (0 == strcasecmp("FATAL", log_level)) {
        return 3;
    }
    return 2;
}

static char const **
get_modules (void)
{
    char const *modules_env = xgetenv("ARIF_RIME_MODULES", NULL);
    if (modules_env == NULL) {
        return NULL;
    }
    size_t modules_len = strlen(modules_env) + 1;

    char *modules_str = xmalloc(sizeof(char) * modules_len);
    memcpy(modules_str, modules_env, modules_len);

    char const **modules = NULL;
    size_t idx = 0;
    for (size_t modules_size = 4; ; *(modules_str++) = '\0') {
        if (modules == NULL || idx == modules_size - 1) {
            modules_size += modules_size / 2;
            modules = xrealloc(modules, sizeof(char const *) * modules_size);
        }
        modules[idx++] = modules_str;

        modules_str = strchr(modules_str, ',');
        if (modules_str == NULL) {
            break;
        }
    }
    modules[idx] = NULL;
    return modules;
}
